src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -56,13 +57,22 @@ private class AggregatingTagger : ForegroundThreadAffinitizedObject, IAccurateTa
                 _owner = owner;
                 _subjectBuffer = subjectBuffer;
 
+                // track context changes
+                _registration = Workspace.GetWorkspaceRegistration(_subjectBuffer.AsTextContainer());
+                _registration.WorkspaceChanged += OnWorkspaceChanged;
+
+                if (_registration.Workspace != null)
+                {
+                    // in unit test, it is possible, we create tagger for buffer which doesn't have associated document
+                    _registration.Workspace.DocumentActiveContextChanged += OnDocumentActiveContextChanged;
+                }
+
                 var document = _subjectBuffer.AsTextContainer().GetOpenDocumentInCurrentContext();
                 _currentDocumentId = document?.Id;
 
                 // Kick off a background task to collect the initial set of diagnostics.
-                var cancellationToken = _initialDiagnosticsCancellationSource.Token;
                 var asyncToken = _owner._listener.BeginAsyncOperation(GetType() + ".GetInitialDiagnostics");
-                var task = Task.Run(() => GetInitialDiagnosticsInBackground(document, cancellationToken), cancellationToken);
+                var task = Task.Run(() => GetInitialDiagnosticsInBackground(document), _shutdownCancellationSource.Token);
```

Comment:
So what happens if the document ID changes while this is running? I see us cancelling it, but I don't see how we'd start the new request for the correct document?

Reply:
turns out this made it work. 
http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/DiagnosticAnalyzerService_IncrementalAnalyzer.cs,58

we lost history so, not sure why that is added. but have feeling it probably related to closed file analysis option. I believe now all documents of same buffer are open or closed together, so I think this won't be needed any more. 

anyway, I am going to remove the above since diagnostic shouldn't care about the context change, and make tagger to deal with context change.